compiler/src/iree/compiler/Dialect/Stream/Transforms/ReplicateGlobalsPerAffinity.cpp

```diff
@@ -0,0 +1,498 @@
+      LDBG() << "\toperand #" << idx << " affinity: ignored";
+      continue;
+    }
+    auto attr = affinityAnalysis.lookupResourceAffinity(value);
+    LDBG() << "\toperand #" << idx << " affinity: " << attr;
+  }
+  for (auto result : affinityOp->getResults()) {
+    SmallVector<IREE::Stream::AffinityAttr> usageAffinities;
+    if (affinityAnalysis.tryLookupResourceUsageAffinity(result,
+                                                        usageAffinities)) {
+      LDBG_OS([&](raw_ostream &os) {
+        os << "\tresult type: " << result.getType() << " affinities: [";
+        llvm::interleaveComma(usageAffinities, os);
+        os << "]";
+      });
+    } else {
+      LDBG_OS([&](raw_ostream &os) {
+        os << "\tresult type: " << result.getType() << " affinities: failed";
+      });
+    }
+  }
+}
+
+void ReplicateGlobalsPerAffinityPass::runOnOperation() {
+  mlir::ModuleOp moduleOp = getOperation();
+  IREE::Stream::AffinityAnalysis affinityAnalysis(moduleOp);
+  if (failed(affinityAnalysis.run())) {
+    LDBG() << "failed on running affinity analysis";
+    return;
+  }
+
+  Explorer explorer(moduleOp.getOperation(), TraversalAction::RECURSE);
+  explorer.setOpInterfaceAction<mlir::FunctionOpInterface>(
+      TraversalAction::RECURSE);
+  explorer.setDialectAction<mlir::scf::SCFDialect>(TraversalAction::RECURSE);
+  explorer.setDialectAction<IREE::Flow::FlowDialect>(TraversalAction::RECURSE);
+  explorer.setDialectAction<IREE::Util::UtilDialect>(TraversalAction::RECURSE);
+  explorer.initialize();
+
+  SetVector<Operation *> worklist;
+  explorer.forEachGlobal([&](const Explorer::GlobalInfo *globalInfo) {
+    if (globalInfo->isIndirect || globalInfo->op.isGlobalMutable()) {
+      return;
+    }
+    // Scalars are expected to be used on host, so we do not need to replicate
+    // them.
+    if (globalInfo->op.getGlobalType().isIntOrIndexOrFloat()) {
+      return;
+    }
+    worklist.insert(globalInfo->op);
+  });
+
+  // List of operands that the source is from global ops.
+  //   [Operation, [(operand number, maybe global op)]].
+  // If the global op is null, it means that the operand is not directly from a
+  // load op. It is important to track the global op, so we can support cross
+  // functions/branches cases.
+  DenseMap<Operation *,
+           SmallVector<std::tuple<OpOperand *, IREE::Util::GlobalOpInterface>>>
+      opToGlobalUseMap;
+  OpOperandAffinityStateMap opOperandAffinityStateMap;
+  SymbolTable symbolTable(moduleOp);
+  while (!worklist.empty()) {
+    Operation *currentOp = *worklist.begin();
+    LDBG() << "processing op: " << *currentOp;
+    worklist.erase(worklist.begin());
+    TypeSwitch<Operation *>(currentOp)
+        .Case<IREE::Util::GlobalOpInterface>([&](auto globalOp) {
+          const Explorer::GlobalInfo *globalInfo =
+              explorer.getGlobalInfo(globalOp);
+          for (auto loadOp : globalInfo->getLoads()) {
```

Comment:
Why only loads and not stores?

Reply:
We filter with immutable globals, so there are only loads. So we only have to handle loads.

https://github.com/iree-org/iree/blob/a93cdb0c18755e76771521e67ce055a07ebecd72/compiler/src/iree/compiler/Dialect/Util/IR/UtilInterfaces.td#L279-L292